lib/engine/step/tokener.rb

```diff
@@ -72,17 +72,25 @@ def adjust_token_price_ability!(entity, token, hex)
         end
 
         entity.abilities(:token) do |ability, _|
-          next unless ability.hexes.include?(hex.id)
+          next if ability.hexes.any? && !ability.hexes.include?(hex.id)
 
           # check if this is correct or should be a corporation
           token = Engine::Token.new(entity) if ability.extra
           token.price = ability.teleport_price if ability.teleport_price
-          token.price = ability.price if @game.graph.reachable_hexes(entity)[hex]
+          token.price = new_token_price(token, ability) if @game.graph.reachable_hexes(entity)[hex]
           return [token, ability]
         end
 
         [token, nil]
       end
+
+      private
+
+      def new_token_price(token, ability)
+        return ability.price if ability.price
```

Comment:
shouldn't this just be abstracted in the ability class?

in class ability, you can have the price reflect the discount

Reply:
Have now replaced price attribute with price method in ability.